cmd/kubeadm/app/cmd/init.go

```diff
@@ -296,6 +296,13 @@ func (i *Init) Run(out io.Writer) error {
 		return err
 	}
 
+	// Upload currently used configuration to the cluster
+	// Note: This is done right in the beginning of cluster initialization; as we might want to make other phases
```

Comment:
This seems fine, but is there a reason we don't want this to occur even before MarkMaster?

Reply:
It feels much safer to wait for the kubelet to register properly before attempting to put things into the API.
I don't think there is gonna be something that MarkMaster needs from the config anyway (and I can live with requiring the phase user to pass that in such a case (only the master name is required now)

Sounds reasonable so this can be lgtm'd?

Reply:
> It feels much safer to wait for the kubelet to register properly before attempting to put things into the API.

Honest question -- why?

But this does answer the crux of my question, which is that we don't actually care about the master being properly marked before we upload the configuration, we're actually just relying on the fact that MarkMaster will wait until the master kubelet is registered. That was the part that was unclear to me, since the comment here claims this is being done "right in the beginning of cluster initialization" but actually happens second. I think that's worth making clear in the comment.

Reply:
fair enough; you've convinced me :smile:
updated to make it the first thing _for real_